packages/flutter/lib/src/widgets/single_child_scroll_view.dart

```diff
@@ -0,0 +1,313 @@
+    switch (scrollDirection) {
+      case Axis.horizontal:
+        return AxisDirection.right;
+      case Axis.vertical:
+        return AxisDirection.down;
+    }
+    return null;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final ScrollBehavior2 scrollBehavior = new _SingleChildScrollViewBehavior(
+      Scrollable2.getScrollBehavior(context));
+    return new Scrollable2(
+      axisDirection: _getDirection(context),
+      initialScrollOffset: initialScrollOffset,
+      scrollBehavior: scrollBehavior,
+      children: child == null ? const <Widget>[] : <Widget>[ child ],
+    );
+  }
+}
+
+class _SingleChildScrollViewBehavior extends ScrollBehavior2Proxy {
+  _SingleChildScrollViewBehavior(ScrollBehavior2 parent) : super(parent);
+
+  @override
+  Widget createViewport({
+    Key key,
+    AxisDirection axisDirection: AxisDirection.down,
+    double anchor: 0.0,
+    ViewportOffset offset,
+    Key center,
+    List<Widget> children: const <Widget>[],
+  }) {
+    assert(anchor == 0.0);
+    assert(center == null);
+    assert(children.length <= 1);
+    return new _SingleChildViewport(
+      key: key,
+      axisDirection: axisDirection,
+      offset: offset,
+      child: children.isEmpty ? null : children[0],
+    );
+  }
+}
+
+class _SingleChildViewport extends SingleChildRenderObjectWidget {
+  _SingleChildViewport({
+    Key key,
+    this.axisDirection: AxisDirection.down,
+    this.offset,
+    Widget child,
+  }) : super(key: key, child: child) {
+    assert(axisDirection != null);
+  }
+
+  final AxisDirection axisDirection;
+  final ViewportOffset offset;
+
+  @override
+  _RenderSingleChildViewport createRenderObject(BuildContext context) {
+    return new _RenderSingleChildViewport(
+      axisDirection: axisDirection,
+      offset: offset,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderSingleChildViewport renderObject) {
+    // Order dependency: The offset setter reads the axis direction.
+    renderObject
+      ..axisDirection = axisDirection
+      ..offset = offset;
+  }
+}
+
+class _RenderSingleChildViewport extends RenderBox with RenderObjectWithChildMixin<RenderBox> {
+  _RenderSingleChildViewport({
+    AxisDirection axisDirection: AxisDirection.down,
+    ViewportOffset offset,
+    RenderBox child,
+  }) : _axisDirection = axisDirection,
+       _offset = offset ?? new ViewportOffset.zero() {
+    assert(axisDirection != null);
+    this.child = child;
+  }
+
+  AxisDirection get axisDirection => _axisDirection;
+  AxisDirection _axisDirection;
+  set axisDirection(AxisDirection value) {
+    assert(value != null);
+    if (value == _axisDirection)
+      return;
+    _axisDirection = value;
+    markNeedsLayout();
+  }
+
+  Axis get axis => axisDirectionToAxis(axisDirection);
+
+  ViewportOffset get offset => _offset;
+  ViewportOffset _offset;
+  set offset(ViewportOffset value) {
+    assert(value != null);
```

Comment:
This doesn't work (I'm fixing it in my block code).
RenderViewport2 now requires a non-null ViewportOffset always (even in the constructor).